### ISSUE TITLE
[PSyData] Dependency Analaysis Improvements

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#641]
PSyKE uses the dependency analysis to decide which variables to store. At this stages it will pick up some variables that are not (strictly) necessary, e.g.:
- loop variables
- temporary scalar variables
Technically those variables could be checked, but it is kind of useless (loop variables), and potentially leading to false positives (temporary scalars when a loop is threaded).

For PSyData in particularly: We should give the user an option to add/remove variables to the automatic list

**Comment by TeranIvy** [comment]
I guess it also depends on what needs to be written out/read in and it may differ between APIs. LFRic for instance has variables defined by API rules that are added to kernel/built-in call argument list. I am not sure about temporary variables, though.

**Comment by hiker** [comment]
Yes, that's to be expected. In gocean this is more obvious (field properties are passed in as explicitly declared arguments). I think it should be easy to change the values written per API (maybe a tiny bit of refactoring, I'll see when I tackle this ticket). The idea is that API-specific implementation can provide whatever they need, and the runtime library provides the functions to write (and read) the required variable types.

**Comment by hiker** [comment]
An additional issue that needs to be support (and might require its own ticket): handling conditional statements. Example:
```
              do jj=1, 10
                 do ji=1, 10
                    b = t(ji, jj) - 2
                    s(ji, jj )=t(ji, jj)+b
                 enddo
              enddo
              do jj=1, 10
                 do ji=1, 10
                    if (t(ji,jj) > 0) then
                        b = sqrt(t(ji, jj))
                    endif
                    s(ji, jj)=t(ji, jj)+b
                 enddo
              enddo
```
These loops cannot be fused in general (considering ``t(1,1)<0`` --> it uses the previous value for ``b``) - but since the current variable access detection ignores the if statement, and just returns that b is written, this can not be detected. X-failing test added in #1075 .
